File: src/lib/converter/map-exporter/map-generate-utils.ts
import chalk from 'chalk';

import { distancePerTile, maxGameHeightDiff } from '@/lib/constants';
import { isWowUnit, WowAdt, WowObject } from '@/lib/converter/common/models';
import type { WowObjectManager } from '@/lib/converter/common/wow-object-manager';
import { MapExportConfig, MapExporter } from '@/lib/converter/map-exporter/map-exporter';
import { computeRecommendedTerrainClampPercent } from '@/lib/converter/map-exporter/wc3-converter';
import { Vector3 } from '@/lib/math/common';
import { V3 } from '@/lib/math/vector';

export function computeCreatureExportSteps(uniqueCreatureCount: number): number {
  return uniqueCreatureCount > 0 ? uniqueCreatureCount : 0;
}

export function countUniqueUnitExportsFromManager(wowObjectManager: WowObjectManager): number {
  const displayIds = new Set<number>();
  let count = 0;
  wowObjectManager.iterateObjects((obj) => {
    if (!isWowUnit(obj)) return;
    const displayId = obj.creature.model.CreatureDisplayID;
    if (!displayId || displayIds.has(displayId)) return;
    displayIds.add(displayId);
    count++;
  });
  return count;
}

export function autoChooseClampPercent(
  mapExporter: MapExporter,
  mapExportConfig: MapExportConfig,
  unitScale: number,
): void {
  const unitPos: Vector3[] = [];
  mapExporter.wowObjectManager.iterateObjects((obj, abs) => {
    if (!isWowUnit(obj)) return;
    unitPos.push(abs.position);
  });
  if (unitPos.length === 0) {
    console.log(
      'No units found, cannot auto choose clamp percent. Defaulting to',
      mapExportConfig.terrain.clampPercent.lower,
      mapExportConfig.terrain.clampPercent.upper,
    );
    return;
  }
  unitPos.sort((a, b) => a[2] - b[2]);
  const { ratio, min, max } = computeRecommendedTerrainClampPercent(mapExporter.wowObjectManager.roots);
  let clampDiff = ratio * unitScale;

  const size = V3.sub(max, min);
  const ratioZ = maxGameHeightDiff / (size[2] * clampDiff);
  const width = size[0] * ratioZ / distancePerTile;
  const height = size[1] * ratioZ / distancePerTile;

  const w4 = Math.ceil(width / 4) * 4;
  const h4 = Math.ceil(height / 4) * 4;
  clampDiff *= Math.max(1, w4 / 480, h4 / 480);

  const unitPosRatio = unitPos.map((pos) => (pos[2] - min[2]) / (max[2] - min[2]));

  let bestLowerPercent = 0;
  let bestUpperPercent = ratio;
  let maxCount = 0;
  const lower = mapExportConfig.terrain.clampPercent.lower;
  const upper = mapExportConfig.terrain.clampPercent.upper;
  if (upper - lower <= clampDiff) {
    console.log('Map terrain clamp is already within the recommended range, skipping auto choose.');
    return;
  }

  for (let lowerPercent = lower; lowerPercent <= upper - clampDiff; lowerPercent += 0.01) {
    const upperPercent = lowerPercent + clampDiff;
    const count = unitPosRatio.filter((r) => r >= lowerPercent && r <= upperPercent).length;
    if (count > maxCount) {
      maxCount = count;
      bestLowerPercent = lowerPercent;
      bestUpperPercent = upperPercent;
    }
  }
  mapExportConfig.terrain.clampPercent.lower = bestLowerPercent;
  mapExportConfig.terrain.clampPercent.upper = bestUpperPercent;
  const leftOutBelow = unitPosRatio.filter((r) => r < bestLowerPercent).length;
  const leftOutAbove = unitPosRatio.filter((r) => r > bestUpperPercent).length;
  const leftOut = leftOutBelow + leftOutAbove;
  const remaining = unitPosRatio.length - leftOut;
  console.log(`Chosen clamp percent: ${bestLowerPercent} - ${bestUpperPercent} (${remaining} units remaining)`);
  console.log(`Left out units: ${leftOut} (${leftOutBelow} below, ${leftOutAbove} above)`);
}

export function pruneDepth(mapExporter: MapExporter, depth: number): void {
  const wowObjectManager = mapExporter.wowObjectManager;
  if (!wowObjectManager) return;

  if (depth >= 3) return;

  const nextRoots: WowObject[] = [];
  const nextObjects = new Map<string, WowObject>();
  const nextDoodads: WowObject[] = [];
  const nextTerrains: WowAdt[] = [];

  const visit = (obj: WowObject, hasWmoAncestor: boolean): WowObject | null => {
    const currentHasWmoAncestor = hasWmoAncestor || obj.type === 'wmo';

    let keep = true;
    if (depth === 1) {
      keep = obj.type === 'adt' || obj.type === 'unit';
    } else if (depth === 2) {
      if (obj.type === 'm2' || obj.type === 'gobj') {
        keep = !currentHasWmoAncestor;
      }
    }

    if (!keep) return null;

    const clone: WowObject = { ...obj, children: [] };
    nextObjects.set(clone.id, clone);

    if (clone.type === 'adt') {
      nextTerrains.push(clone as WowAdt);
    } else if (clone.type !== 'unit') {
      nextDoodads.push(clone);
    }

    for (const child of obj.children) {
      const prunedChild = visit(child, currentHasWmoAncestor);
      if (prunedChild) clone.children.push(prunedChild);
    }

    return clone;
  };

  for (const root of wowObjectManager.roots) {
    const newRoot = visit(root, false);
    if (newRoot) nextRoots.push(newRoot);
  }

  if (depth === 1) {
    const hasAdtRoot = nextRoots.some((root) => root.type === 'adt');
    wowObjectManager.roots = hasAdtRoot
      ? nextRoots.filter((root) => root.type === 'adt')
      : nextRoots;
  } else {
    wowObjectManager.roots = nextRoots;
  }

  wowObjectManager.objects = nextObjects;
  wowObjectManager.doodads = nextDoodads;
  if (nextTerrains.length > 0) {
    wowObjectManager.terrains = nextTerrains;
  }
}

export function logMapGeneratePhase(label: string): void {
  console.log(chalk.cyan(`[map-generate] ${label}`));
}

File: golang/internal/converter/mapexporter/map_generate_utils.go
package mapexporter

import (
	"log"
	"sort"

	"github.com/pqhuy98/wow-converter/internal/config"
	"github.com/pqhuy98/wow-converter/internal/converter/common"
	"github.com/pqhuy98/wow-converter/internal/converter/convertlog"
	"github.com/pqhuy98/wow-converter/internal/math"
)

const CreatureExportConcurrency = 5

// ComputeCreatureExportSteps counts creature export progress steps.
func ComputeCreatureExportSteps(uniqueCreatureCount int) int {
	if uniqueCreatureCount > 0 {
		return uniqueCreatureCount
	}
	return 0
}

// CountUniqueUnitExportsFromManager counts unique unit display IDs.
func CountUniqueUnitExportsFromManager(m *common.WowObjectManager) int {
	displayIDs := map[int]struct{}{}
	count := 0
	m.IterateObjects(func(obj *common.WowObject, _ common.ObjectAbsolute) {
		c := common.ObjectCreature(obj)
		if !common.IsWowUnit(obj) || c == nil {
			return
		}
		displayID := c.Model.CreatureDisplayID
		if displayID == 0 {
			return
		}
		if _, ok := displayIDs[displayID]; ok {
			return
		}
		displayIDs[displayID] = struct{}{}
		count++
	})
	return count
}

// AutoChooseClampPercent picks terrain clamp to fit most units.
func AutoChooseClampPercent(exporter *MapExporter, unitScale float64) {
	mapCfg := exporter.MapExportConfig
	if mapCfg == nil {
		return
	}
	var unitPos []math.Vector3
	exporter.WowObjectManager.IterateObjects(func(obj *common.WowObject, abs common.ObjectAbsolute) {
		if common.IsWowUnit(obj) {
			unitPos = append(unitPos, abs.Position)
		}
	})
	if len(unitPos) == 0 {
		log.Printf("No units found, cannot auto choose clamp percent. Defaulting to %.2f %.2f",
			mapCfg.Terrain.ClampPercent.Lower, mapCfg.Terrain.ClampPercent.Upper)
		return
	}

	sort.Slice(unitPos, func(i, j int) bool { return unitPos[i][2] < unitPos[j][2] })
	ratio, min, max := ComputeRecommendedTerrainClampPercent(exporter.WowObjectManager.Roots)
	clampDiff := ratio * unitScale
	size := math.V3Sub(max, min)
	ratioZ := config.MaxGameHeightDiff() / (size[2] * clampDiff)
	width := size[0] * ratioZ / config.DistancePerTile
	height := size[1] * ratioZ / config.DistancePerTile
	w4 := ceilDiv4(width)
	h4 := ceilDiv4(height)
	clampDiff *= maxFloat(1, w4/480, h4/480)

	lower := mapCfg.Terrain.ClampPercent.Lower
	upper := mapCfg.Terrain.ClampPercent.Upper
	if upper-lower <= clampDiff {
		log.Println("Map terrain clamp is already within the recommended range, skipping auto choose.")
		return
	}

	bestLower, bestUpper := lower, lower+clampDiff
	maxCount := 0
	unitPosRatio := make([]float64, len(unitPos))
	for i, pos := range unitPos {
		unitPosRatio[i] = (pos[2] - min[2]) / (max[2] - min[2])
	}
	for lowerPercent := lower; lowerPercent <= upper-clampDiff; lowerPercent += 0.01 {
		upperPercent := lowerPercent + clampDiff
		count := 0
		for _, r := range unitPosRatio {
			if r >= lowerPercent && r <= upperPercent {
				count++
			}
		}
		if count > maxCount {
			maxCount = count
			bestLower = lowerPercent
			bestUpper = upperPercent
		}
	}
	mapCfg.Terrain.ClampPercent.Lower = bestLower
	mapCfg.Terrain.ClampPercent.Upper = bestUpper
	leftOutBelow, leftOutAbove := 0, 0
	for _, r := range unitPosRatio {
		if r < bestLower {
			leftOutBelow++
		}
		if r > bestUpper {
			leftOutAbove++
		}
	}
	leftOut := leftOutBelow + leftOutAbove
	remaining := len(unitPosRatio) - leftOut
	log.Printf("Chosen clamp percent: %.2f - %.2f (%d units remaining)", bestLower, bestUpper, remaining)
	log.Printf("Left out units: %d (%d below, %d above)", leftOut, leftOutBelow, leftOutAbove)
}

// PruneDepth limits object tree depth (1=adt, 2=+top m2, 3=+interiors).
func PruneDepth(exporter *MapExporter, depth int) {
	if exporter.WowObjectManager == nil || depth >= 3 {
		return
	}
	m := exporter.WowObjectManager
	nextRoots := []*common.WowObject{}
	nextObjects := map[string]*common.WowObject{}
	nextDoodads := []*common.WowObject{}
	nextTerrains := []*common.WowObject{}

	var visit func(obj *common.WowObject, hasWmoAncestor bool) *common.WowObject
	visit = func(obj *common.WowObject, hasWmoAncestor bool) *common.WowObject {
		currentHasWmo := hasWmoAncestor || obj.Type == common.WowObjectWMO
		keep := true
		switch depth {
		case 1:
			keep = obj.Type == common.WowObjectADT || obj.Type == common.WowObjectUnit
		case 2:
			if obj.Type == common.WowObjectM2 || obj.Type == common.WowObjectGobj {
				keep = !currentHasWmo
			}
		}
		if !keep {
			return nil
		}
		clone := *obj
		clone.Children = nil
		nextObjects[clone.ID] = &clone
		if clone.Type == common.WowObjectADT {
			nextTerrains = append(nextTerrains, &clone)
		} else if clone.Type != common.WowObjectUnit {
			nextDoodads = append(nextDoodads, &clone)
		}
		for _, child := range obj.Children {
			if pruned := visit(child, currentHasWmo); pruned != nil {
				clone.Children = append(clone.Children, pruned)
			}
		}
		return &clone
	}

	for _, root := range m.Roots {
		if newRoot := visit(root, false); newRoot != nil {
			nextRoots = append(nextRoots, newRoot)
		}
	}
	if depth == 1 {
		hasAdt := false
		for _, r := range nextRoots {
			if r.Type == common.WowObjectADT {
				hasAdt = true
				break
			}
		}
		if hasAdt {
			filtered := nextRoots[:0]
			for _, r := range nextRoots {
				if r.Type == common.WowObjectADT {
					filtered = append(filtered, r)
				}
			}
			nextRoots = filtered
		}
	}
	m.Roots = nextRoots
	m.Objects = nextObjects
	m.Doodads = nextDoodads
	if len(nextTerrains) > 0 {
		m.Terrains = nextTerrains
	}
}

// LogMapGeneratePhase logs a map generation phase label.
func LogMapGeneratePhase(label string) {
	convertlog.MapGenerate(label)
}

func ceilDiv4(v float64) float64 {
	return float64((int(v) + 3) / 4 * 4)
}

func maxFloat(a float64, rest ...float64) float64 {
	m := a
	for _, v := range rest {
		if v > m {
			m = v
		}
	}
	return m
}
